library/Icingadb/View/BaseHostAndServiceRenderer.php

```diff
@@ -0,0 +1,263 @@
+use ipl\Html\HtmlString;
+use ipl\Html\Text;
+use ipl\Html\ValidHtml;
+use ipl\I18n\Translation;
+use ipl\Web\Common\ItemRenderer;
+use ipl\Web\Widget\EmptyState;
+use ipl\Web\Widget\Icon;
+use ipl\Web\Widget\StateBall;
+use ipl\Web\Widget\TimeSince;
+
+/** @implements ItemRenderer<Host|Service> */
+abstract class BaseHostAndServiceRenderer implements ItemRenderer
+{
+    use Translation;
+
+    abstract protected function createSubject($item, string $layout): ValidHtml;
+
+    public function assembleVisual($item, HtmlDocument $visual, string $layout): void
+    {
+        if ($layout === 'header') {
+            if ($item->state->state_type === 'soft') {
+                $stateType = 'soft_state';
+                $previousStateType = 'previous_soft_state';
+
+                if ($item->state->previous_soft_state === 0) {
+                    $previousStateType = 'hard_state';
+                }
+            } else {
+                $stateType = 'hard_state';
+                $previousStateType = 'previous_hard_state';
+
+                if ($item->state->hard_state === $item->state->previous_hard_state) {
+                    $previousStateType = 'previous_soft_state';
+                }
+            }
+
+            if ($item instanceof Host) {
+                $state = HostStates::text($item->state->$stateType);
+                $previousState = HostStates::text($item->state->$previousStateType);
+            } else {
+                $state = ServiceStates::text($item->state->$stateType);
+                $previousState = ServiceStates::text($item->state->$previousStateType);
+            }
+
+            $stateChange = new StateChange($state, $previousState);
+            if ($stateType === 'soft_state') {
+                $stateChange->setCurrentStateBallSize(StateBall::SIZE_MEDIUM_LARGE);
+            }
+
+            if ($previousStateType === 'previous_soft_state') {
+                $stateChange->setPreviousStateBallSize(StateBall::SIZE_MEDIUM_LARGE);
+                if ($stateType === 'soft_state') {
+                    $visual->getAttributes()->add('class', 'small-state-change');
+                }
+            }
+
+            $stateChange->setIcon($item->state->getIcon());
+            $stateChange->setHandled(
+                $item->state->is_problem && ($item->state->is_handled || ! $item->state->is_reachable)
+            );
+
+            $visual->addHtml($stateChange);
+
+            return;
+        }
+
+        $ballSize = $layout === 'minimal' ? StateBall::SIZE_BIG : StateBall::SIZE_LARGE;
+
+        $stateBall = new StateBall($item->state->getStateText(), $ballSize);
+        $stateBall->add($item->state->getIcon());
+        if ($item->state->is_problem && ($item->state->is_handled || ! $item->state->is_reachable)) {
+            $stateBall->getAttributes()->add('class', 'handled');
+        }
+
+        $visual->addHtml($stateBall);
+        if ($layout !== 'minimal' && $item->state->state_type === 'soft') {
```

Comment:
Thank you! Though, please remember to also drop the CSS rule previously used to hide it.